haskell/cabal.bzl

```diff
@@ -243,3 +243,225 @@ However, using a plain `haskell_library` sometimes leads to better
+            [src, _, dest] = tokens
+            if src in all_packages and dest in all_packages:
+                dependencies[src].append(dest)
+    return (dependencies, transitive_unpacked_sdists)
+
+def _stack_install_impl(repository_ctx):
+    packages = repository_ctx.attr.packages
+    non_core_packages = [
+        package
+        for package in packages
+        if package not in _CORE_PACKAGES
+    ]
+    versioned_packages = []
+    unversioned_packages = []
+    for package in non_core_packages:
+        if package.rpartition("-")[2].replace(".", "").isdigit():
+            versioned_packages.append(package)
+        else:
+            unversioned_packages.append(package)
+    (dependencies, transitive_unpacked_sdists) = _compute_dependency_graph(
+        repository_ctx,
+        versioned_packages,
+        unversioned_packages,
+    )
+
+    # Write out the dependency graph as a BUILD file.
+    build_file_builder = []
+    build_file_builder.append("""
+load("@io_tweag_rules_haskell//haskell:cabal.bzl", "haskell_cabal_library")
+load("@io_tweag_rules_haskell//haskell:haskell.bzl", "haskell_toolchain_library")
+""")
+    extra_deps = [
+        "@{}//{}:{}".format(label.workspace_name, label.package, label.name)
+        for label in repository_ctx.attr.deps
+    ]
+    for package in _CORE_PACKAGES:
+        if package in packages:
+            visibility = ["//visibility:public"]
+        else:
+            visibility = ["//visibility:private"]
+        build_file_builder.append(
+            """
+haskell_toolchain_library(name = "{name}", visibility = {visibility})
+""".format(name = package, visibility = visibility),
+        )
+    for package in transitive_unpacked_sdists:
+        unversioned_package = _chop_version(package)
+        if unversioned_package in _CORE_PACKAGES:
+            continue
+        if unversioned_package in unversioned_packages or package in versioned_packages:
+            visibility = ["//visibility:public"]
+        else:
+            visibility = ["//visibility:private"]
+        build_file_builder.append(
+            """
+haskell_cabal_library(
+    name = "{name}",
+    srcs = glob(["{dir}/**"]),
+    deps = {deps},
+    visibility = {visibility},
+)
+""".format(
+                name = package,
+                dir = package,
+                deps = dependencies[unversioned_package] + extra_deps,
+                testonly = repository_ctx.attr.testonly,
+                visibility = visibility,
+            ),
+        )
+        build_file_builder.append(
+            """alias(name = "{name}", actual = ":{actual}", visibility = {visibility})""".format(
+                name = unversioned_package,
+                actual = package,
+                visibility = visibility,
+            ),
+        )
+    build_file_content = "\n".join(build_file_builder)
+    repository_ctx.file("BUILD.bazel", build_file_content, executable = False)
+
+stack_install = repository_rule(
```

Comment:
Comparing to Hazel, here are a few things that come to mind which `stack_install` and `haskell_cabal_library` don't seem to offer, yet. Those might be good future extensions.
- Override individual packages with custom Bazel builds. Probably easiest by injecting a mapping from package name to Bazel targets.
- Stack's `extra-deps` feature to override a package version or just it's source repository. Also to add additional Hackage packages, or packages from other sources.
- Apply source patches. This is very useful if a package just requires minor adjustment to build within Bazel. E.g. file paths to embedded data files. In such cases a full fork imposes a lot of overhead.
- Overriding cabal flags. E.g. `integer-simpl` vs. `integer-gmp`.
- Binary targets. As I understand `haskell_cabal_library` does currently not expose any binary targets. I suspect this has two consequences. A) Users cannot use e.g. a Stackage provided `c2hs`. B) Stackage packages that depend on other Stackage packages' binaries, e.g. `happy`, will probably not work.

Reply:
> Override individual packages with custom Bazel builds.

Not currently possible, but it's possible to override the version at least.

> Stack's extra-deps feature to override a package version or just it's source repository.

See above. Not possible to override with one's own source repository. Good future feature.

> Also to add additional Hackage packages, or packages from other sources.

Outside of the cases listed above, these wouldn't be part of the transitive closure, right? So don't need to be part of the `stackage` external repository?

> Apply source patches ... Overriding cabal flags ... Binary targets

Agreed. Let's create tickets about those.

Reply:
> Stack's extra-deps feature [...]

I haven't tested any of these new features yet, however in practice one can make a custom snapshot file, host it remotely, and point `stack` to it (e.g. `stack build --resolver=<link_to_snapshot.>`).

Depending on how `rules_haskell` and `stack` interact here, this could be a neat workaround for `extra-deps` at the moment.

Reply:
Hm, so you're saying #910 could be unnecessary? It hasn't been merged yet. Not too late to block it. :)

Reply:
Unfortunately I can't say _for sure_, since I've been unable to get the quickstart instructions for `rules_haskell` working on macOS; I'm basing this purely off of my knowledge of how `stack` treats resolvers.

I do think that having something like #910 would probably be useful, regardless of this, though. I just wanted to suggest it just in case it actually did work, and could help unblock someone in the meantime 😄

Reply:
@jkachmar You can use `rules_haskell` master for now, on MacOS there was a minor bugfix, see https://github.com/tweag/rules_haskell/pull/893

I’m gonna push a point-release later today.

Reply:
There's a singular point missing in our CI matrix currently, which I guess is the point you're in: macOS + ghc bindists (as opposed to GHC-from-Nixpkgs).

Reply:
https://github.com/tweag/rules_haskell/releases/tag/v0.9.1